lib/ddtrace/contrib/sinatra/tracer.rb

```diff
@@ -32,53 +36,115 @@ def route(verb, action, *)
-          ::Sinatra::Base.module_eval do
-            def render(engine, data, *)
-              output = ''
-              tracer = Datadog.configuration[:sinatra][:tracer]
-              if tracer.enabled
-                tracer.trace(Ext::SPAN_RENDER_TEMPLATE, span_type: Datadog::Ext::HTTP::TEMPLATE) do |span|
-                  # If data is a string, it is a literal template and we don't
-                  # want to record it.
-                  span.set_tag(Ext::TAG_TEMPLATE_NAME, data) if data.is_a? Symbol
-
-                  # Measure service stats
-                  Contrib::Analytics.set_measured(span)
-
-                  output = super
-                end
-              else
-                output = super
-              end
+          app.use TracerMiddleware
 
-              output
-            end
-          end
+          app.after do
+            configuration = Datadog.configuration[:sinatra]
+            return unless configuration[:tracer].enabled
+            return if Sinatra::Env.middleware_traced?(env)
 
-          app.use TracerMiddleware
+            span = Sinatra::Tracer.fetch_middleware_span!(env, configuration)
 
-          app.before do
-            return unless Datadog.configuration[:sinatra][:tracer].enabled
+            route = if defined?(@datadog_route)
+                      @datadog_route
+                    else
+                      # Fallback in case no routes have matched
+                      request.path
+                    end
+
+            span.resource = "#{request.request_method} #{route}"
 
-            span = Sinatra::Env.datadog_span(env)
             span.set_tag(Datadog::Ext::HTTP::URL, request.path)
             span.set_tag(Datadog::Ext::HTTP::METHOD, request.request_method)
+            span.set_tag(Ext::TAG_APP_NAME, app.settings.name)
+            span.set_tag(Ext::TAG_ROUTE_PATH, route)
+            if request.script_name && !request.script_name.empty?
+              span.set_tag(Ext::TAG_SCRIPT_NAME, request.script_name)
+            end
+
+            span.set_tag(Datadog::Ext::HTTP::STATUS_CODE, response.status)
+            span.set_error(env['sinatra.error']) if response.server_error?
+
+            Sinatra::Env.set_middleware_traced(env, true)
           end
+        end
 
-          app.after do
-            return unless Datadog.configuration[:sinatra][:tracer].enabled
+        # Retrieves of initializes the top-most middleware Sinatra span.
+        # This allows us to only create spans for Sinatra middlewares that
+        # have a matching route.
+        #
+        # If we traced all Sinatra middleware apps, we would have a chain
+        # of nested spans that were not responsible for handling this request,
+        # adding little value to users.
+        def self.fetch_middleware_span!(env, configuration)
+          span = Sinatra::Env.datadog_span(env)
+          return span if span
+
+          tracer = configuration[:tracer]
+          span = tracer.trace(
```

Comment:
Out of curiosity, why auto-create the span on `fetch`?

Is this somehow related to if the stack short-circuits before reaching `route_eval`?

Reply:
Yes, this is the part that create the span when we are sure this is the correct middleware layer to instrument.
It can only happen in two cases:
* When a route matches: we create it right before we instrument the route itself.
* When not route matches: we create it on the `app.after` callback, before we return from the current middleware.